client/client-multi/client-c/tools/kaa_encryption/rsa_key_gen/kaa_rsa_key_gen.c

```diff
@@ -210,76 +198,122 @@ static int sha1_to_base64(uint8_t *key, size_t length, uint8_t *base64, size_t b
+        error = 1;
+        goto exit;
+    }
 
     key_begins = private_key + KAA_RSA_PRIVATE_KEY_LENGTH_MAX - private_key_length;
-    store_key(fd, KAA_RSA_PRIVATE_KEY, sizeof(KAA_RSA_PRIVATE_KEY) - 1, key_begins, private_key_length);
+    if (store_key(fd, KAA_RSA_PRIVATE_KEY, sizeof(KAA_RSA_PRIVATE_KEY) - 1, key_begins, private_key_length)) {
+        error = 1;
+        goto exit;
+    }
 
     size_t sha1_base64_len = 0;
     uint8_t sha1[SHA1_LENGTH];
     unsigned char sha1_base64_buffer[1024];
     memset(sha1_base64_buffer, 0, sizeof(sha1_base64_buffer));
     sha1_from_public_key(keys.public_key, keys.public_key_length, sha1);
 
-    int error = sha1_to_base64(sha1, sizeof(sha1), sha1_base64_buffer, sizeof(sha1_base64_buffer), &sha1_base64_len);
+    error = sha1_to_base64(sha1, sizeof(sha1), sha1_base64_buffer, sizeof(sha1_base64_buffer), &sha1_base64_len);
     if (error) {
         printf("Error while encoding base64");
+        goto exit;
     }
 
-    written = sprintf(buffer, KAA_SHA1_PUB_LEN, sizeof(sha1));
-    fwrite(buffer, written, 1, fd);
+    written = snprintf_check(sizeof(buffer) - 1, snprintf(buffer, sizeof(buffer) - 1, KAA_SHA1_PUB_LEN, sizeof(sha1)));
+    if (fwrite_check(fwrite(buffer, written, 1, fd), 1)) {
+        error = 1;
+        goto exit;
+    }
 
-    written = sprintf(buffer, KAA_SHA1_PUB_BASE64_LEN, sha1_base64_len);
-    fwrite(buffer, written, 1, fd);
+    written = snprintf_check(sizeof(buffer) - 1, snprintf(buffer, sizeof(buffer) - 1, KAA_SHA1_PUB_BASE64_LEN, sha1_base64_len));
+    if (fwrite_check(fwrite(buffer, written, 1, fd), 1)) {
+        error = 1;
+        goto exit;
+    }
 
     error = sha1_store(fd, sha1, sizeof(sha1), sha1_base64_buffer, sha1_base64_len);
     if (error) {
         printf("Can't store sha1\n");
     }
 
-    fwrite(GUARD_ENDIF, sizeof(GUARD_ENDIF) - 1, 1, fd);
-    fclose(fd);
+    if (fputs(GUARD_ENDIF, fd) <= 0) {
+        error = 1;
+    }
 
-    return 0;
+exit:    
+    fclose(fd);
+    return error;
 }
 
-int main()
+int write_rsa_key(mbedtls_pk_context *key, const char *output_file, int mode)
 {
-    keys.public_key_length = KAA_RSA_PUBLIC_KEY_LENGTH_MAX;
-    keys.private_key_length = KAA_RSA_PRIVATE_KEY_LENGTH_MAX;
+    FILE *f;
+    unsigned char output_buf[16000];
+    unsigned char *c = output_buf;
+    size_t len = 0;
+
+    memset(output_buf, 0, 16000);
+    if (mode == PRIVATE_KEY) {
+        if (mbedtls_pk_write_key_pem( key, output_buf, 16000) != 0) {
+            return 1;
+        }
+    } else if (mode == PUBLIC_KEY) {
+        if (mbedtls_pk_write_pubkey_pem(key, output_buf, 16000) != 0) {
+            return 1;
+        }
+    } else {
+        return 1;
+    }
 
-    int error = rsa_keys_create(&pk_context_, keys.public_key, &keys.public_key_length,
-                                keys.private_key, &keys.private_key_length);
-    if (error) {
-        printf("Error: can't generate keys (%i)\n", error);
+    len = strlen((char *)output_buf );
```

Comment:
No
Keys generated in pem format.
They are Base64 encoded ASCII files and contain "-----BEGIN CERTIFICATE-----" and "-----END CERTIFICATE-----" statements.